### ISSUE TITLE
[API Proposal]: Introduce token versions to deal with token hiding in linked CTS

### ISSUE THREAD

**Issue by sakno** [dotnet/runtime#69556]
### Background and motivation

I would like to return back to #22172 discussion. I faced with the same issue when I can't propagate user-supplied cancellation token in _OperationCanceledException_.

### API Proposal

We don't want to introduce a new heavy API and overloads that can impact performance and backward compatibility both. Instead, we can distinguish the order of cancellations and determine which token was canceled first.

```csharp
namespace System.Threading;

public readonly struct CancellationToken
{
   public long Version { get; }
}
```
The number is increasing monotonically for each cancellation happened in the app. So the counter is not local to the CTS. On every instantiation and successful _TryReset_, CTS must increase the counter by 1 which is easily achievable via static volatile variable and `IncrementAndGet` interlocked method. The incremented value is stored internally in CTS.

### API Usage

Now we can easily deal with the example shown by David:
```csharp
var cts1 = new CancellationTokenSource();
var cts2 = new CancellationTokenSource();
var linked = CancellationTokenSource.CreateLinkedTokenSource(cts1.Token, cts2.Token);
try
{
    cts1.Cancel();
    linked.Token.ThrowIfCancellationRequested();
}
catch (OperationCanceledException ex)
{
    var version1 = cts1.Token.Version;
    var version2 = cts2.Token.Version;

    switch ((cts1.Token.IsCancellationRequested, cts2.Token.IsCancellationRequested))
    {
       case (true, true):
        throw new OperationCanceledException(version1 < version2 ? cts1.Token : cts2.Token);
       case (true, false):
        throw new OperationCanceledException(cts1.Token);
       case (false, true):
        throw new OperationCanceledException(cts2.Token);
    }
}
```

### Risks

This proposal requires minimal impact on performance because we need a single interlocked operation for that. Increment happens in CTS constructor and _TryReset_ method. Cancellation procedure itself is not affected.

**Comment by svick** [comment]
Why does it matter which token caused the cancellation? If both are cancelled, isn't it enough to report either one of them?

**Comment by sakno** [comment]
Here is practical example from _HttpClient_ that currently requires workaround:
https://github.com/dotnet/runtime/blob/215b39abf947da7a40b0cb137eab4bceb24ad3e3/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L594-L613

**Comment by stephentoub** [comment]
> Here is practical example from HttpClient

How would your proposal materially change that snippet?

**Comment by stephentoub** [comment]
> But we want to propagate a correct cancellation token to the caller.

How do you define "correct"?  In most of these situations, it doesn't matter which actually triggered the cancellation; if they're both canceled, it's a race condition, either of them could have caused it, and it doesn't matter which is used.

**Comment by sakno** [comment]
@stephentoub , agree, sequence number is not an option here. However, CTS is still able to detect the actual token caused cancellation. We can add extra property to _CancellationToken_ to obtain original token:
```csharp
public readonly struct CancellationToken
{
   public CancellationToken? Parent { get; }
}
```

The property throws _InvalidOperationException_ if the token is not canceled, **null** if there is no parent token caused cancellation and actual token that causes cancellation.
